Added Churros for Economic

MadhuriKadam-GSLab opened the pull request:
## Highlights
* Added the churros test for the Economic element.
## Non Customer Highlights
- POST `/journals/${id}/accounting-years/${year}/vouchers/${Id}/attachments` API is not working from churros and from new UI but it working from old UI and from postman.Hence added the skip for that test.
## Screenshot
![churros1](https://user-images.githubusercontent.com/22440353/36591913-900afd2e-18b9-11e8-8f8a-6b0eeb40699e.JPG)
![churros2](https://user-images.githubusercontent.com/22440353/36591915-9036e3f8-18b9-11e8-95c6-47b770d9bc3b.JPG)
![churros3](https://user-images.githubusercontent.com/22440353/36591912-8fdca88e-18b9-11e8-9ab4-fb226f7c6697.JPG)

## Reference
* [SOBA](https://github.com/cloud-elements/soba/pull/8134)
* [churros-sauce](https://github.com/cloud-elements/churros-sauce/pull/166)
* [RALLY-#${US7380}](https://rally1.rallydev.com/#/144349237612d/detail/userstory/193385144768)

## Closes
* [US7380](https://rally1.rallydev.com/#/144349237612d/detail/userstory/193385144768)

GarimaAgarwal-GSLab commented on src/test/elements/economic/assets/contact.json:
```diff
@@ -0,0 +1,8 @@
+      "phone":"96782219205",
+        "notes":"sss",
+          "emailNotifications":["reminders"]
+
```
Indent this.

GarimaAgarwal-GSLab replied:
Indent this.

GarimaAgarwal-GSLab commented on src/test/elements/economic/assets/products.json:
```diff
@@ -0,0 +1,9 @@
+            "productGroupNumber": 1
+      },
+        "barred":false,
+          "costPrice":50
```
Indent this.

GarimaAgarwal-GSLab replied:
Indent this.

GarimaAgarwal-GSLab commented:
@MadhuriKadam-GSLab , Add SOBA PR Link and CLoses section.

tyler-cloud-elements commented on src/test/elements/economic/accounting-years.js:
```diff
@@ -0,0 +1,55 @@
+       let id,pId;
+       return cloud.get(`${test.api}`)
+         .then(r => {
+           id = r.body[0].id;})
```
You don't really need the braces here, so just `.then(r => id = r.body[0].id)`

MadhuriKadam-GSLab replied:
done

tyler-cloud-elements commented on src/test/elements/economic/accounting-years.js:
```diff
@@ -0,0 +1,55 @@
+         .then(r => cloud.get(`${test.api}/${id}/periods`))
+         .then(r => {
+           pId = r.body[0].periodNumber;
+         })
```
Same as above

MadhuriKadam-GSLab replied:
done

tyler-cloud-elements commented on src/test/elements/economic/accounting-years.js:
```diff
@@ -0,0 +1,55 @@
+         .then(r => cloud.get(`${test.api}/${id}/periods/${pId}`))
+         .then(r => cloud.get(`${test.api}/${id}/periods/${pId}/entries`))
+         .then(r => cloud.withOptions({ qs: { where: `date = '2016-12-31'` } }).get(`${test.api}/${id}/periods/${pId}/entries`))
+         .then(r => cloud.withOptions({ qs: { where: `accountNumber = 1010` } }).get(`${test.api}/${id}/periods/${pId}/entries`));
```
You have a lot of `where` tests here, which is good, but it'd be more effective if we had some validation on these. It might be worth creating a common validation method so you can cut down on some of this

MadhuriKadam-GSLab replied:
Added the validation per test not added the function

tyler-cloud-elements commented on src/test/elements/economic/accounting-years.js:
```diff
@@ -0,0 +1,55 @@
+        .should.return200OnGet();
+
+
+     it('should support S for /accounting-years/:year/entires', () => {
```
`entries` not `entires`

tyler-cloud-elements commented on src/test/elements/economic/accounting-years.js:
```diff
@@ -0,0 +1,55 @@
+       expect(validValues.length).to.equal(r.body.length);
+     })
+     .should.return200OnGet();
+     test
```
Format file plz

tyler-cloud-elements commented on src/test/elements/economic/customer.js:
```diff
@@ -0,0 +1,54 @@
+const tools = require('core/tools');
+const chakram = require('chakram');
+const cloud = require('core/cloud');
+const expect = require('chakram').expect;
```
As you already required chakram, you can just do `chakram.expect` here

tyler-cloud-elements commented on src/test/elements/economic/customer.js:
```diff
@@ -0,0 +1,54 @@
+      .then(r => cloud.get(`${test.api}/${customerId}/delivery-locations/${id}`))
+      .then(r => cloud.delete(`${test.api}/${customerId}/delivery-locations/${id}`));
+  });
+  after(() => cloud.delete(`${test.api}/${customerId}`));
```
Can you move this next to the `before`

tyler-cloud-elements commented on src/test/elements/economic/customer.js:
```diff
@@ -0,0 +1,54 @@
+    let id;
+    return cloud.post(`${test.api}/${customerId}/contacts`, contactPayload)
+      .then(r => {
+        id = r.body.customerContactNumber;
```
If you add this to the transformation, you can do 
``` 
 it(`should allow CRUDS for ${test.api}/:id/contacts `, () => cloud.cruds(`${test.api}/${customerId}/contacts `, contactPayload));
```

MadhuriKadam-GSLab replied:
done

tyler-cloud-elements commented on src/test/elements/economic/customer.js:
```diff
@@ -0,0 +1,54 @@
+    let id;
+    return cloud.post(`${test.api}/${customerId}/delivery-locations`, deliveryLocationPayload)
+      .then(r => {
+        id = r.body.deliveryLocationNumber;
```
If you add this to the transformation, you can do 
``` 
 it(`should allow CRUDS for ${test.api}/:id/delivery-locations `, () => cloud.cruds(`${test.api}/${customerId}/delivery-locations `, deliveryLocationPayload));
```

MadhuriKadam-GSLab replied:
done

tyler-cloud-elements commented on src/test/elements/economic/customerGroups.js:
```diff
@@ -0,0 +1,36 @@
+const tools = require('core/tools');
+const chakram = require('chakram');
+const cloud = require('core/cloud');
+const expect = require('chakram').expect;
```
As you already required chakram, you can just do `chakram.expect` here

MadhuriKadam-GSLab replied:
done

tyler-cloud-elements commented on src/test/elements/economic/customerGroups.js:
```diff
@@ -0,0 +1,36 @@
+      }
+    }
+  };
+  payload.customerGroupNumber=tools.randomInt();
```
Can you add this next to the `const payload`

MadhuriKadam-GSLab replied:
done

tyler-cloud-elements commented on src/test/elements/economic/invoice-drafts.js:
```diff
@@ -0,0 +1,37 @@
+const tools = require('core/tools');
+const chakram = require('chakram');
+const cloud = require('core/cloud');
+const expect = require('chakram').expect;
```
As you already required chakram, you can just do `chakram.expect` here

MadhuriKadam-GSLab replied:
done

tyler-cloud-elements commented on src/test/elements/economic/invoice-drafts.js:
```diff
@@ -0,0 +1,37 @@
+    .then(r => cloud.post('/hubs/erp/customers', customerPayload))
+    .then(r => {customerId = r.body.id ;
+              customerId=parseInt(customerId);
+    })
```
Can you format this ?

tyler-cloud-elements commented on src/test/elements/economic/invoice-drafts.js:
```diff
@@ -0,0 +1,37 @@
+    })
+    .then(r => payload.customer.customerNumber = customerId)
+    .then(r => payload.lines[0].product.productNumber = productId)
+     );
```
Can you add an `after` to clean up these customers/products ?

MadhuriKadam-GSLab replied:
done

tyler-cloud-elements commented on src/test/elements/economic/invoices-totals.js:
```diff
@@ -0,0 +1,26 @@
+    .should.return200OnGet();
+  test
+    .withOptions({ qs: { where: `type = 'booked' and subType = 'unpaid' and due = 'not-overdue' ` } }).withName('should support Ceql type,due and sub-type search')
+    .should.return200OnGet();
```
Glad that you're adding all the variations here, but is there anything coming back that you can validate the query with ?

MadhuriKadam-GSLab replied:
@tyler-cloud-elements Updated the test

tyler-cloud-elements commented on src/test/elements/economic/invoices.js:
```diff
@@ -0,0 +1,46 @@
+     .should.return200OnGet();
+     it('should support PUT  for /invoices/:id/', () => {
+       return cloud.put(`${test.api}/${Id}`);
+
```
Can you format this? Didn't notice this the first time through

tyler-cloud-elements commented on src/test/elements/economic/invoices.js:
```diff
@@ -0,0 +1,46 @@
+    .then(r => {
+      Id = r.body.id;
+      Id=parseInt(Id);
+    } )
```
Same as above, format and add an `after` to clean up `products`/`customers`

MadhuriKadam-GSLab replied:
@tyler-cloud-elements Adding skip : true for this API because, we can only edit\delete the draft invoices but once we convert the draft invoice to sent we can not delete the invoices and hence we can not delete product which is in use.  In this test, we are converting draft invoice to sent invoice.

tyler-cloud-elements commented on src/test/elements/economic/journals.js:
```diff
@@ -0,0 +1,46 @@
+           year = r.body.accountingYear.year;
+         })
+         .then(r => cloud.get(`/hubs/erp/journals/${id}/accounting-years/${year}/vouchers/${vId}`))
+         .then(r => cloud.withOptions({ headers: { "Content-Type": "multipart/form-data" }}).postFile(`/hubs/erp/journals/${id}/accounting-years/${year}/vouchers/${vId}/attachments`,jpgFile))
```
You shouldn't need to specify this `Content-Type`, `postFile` should handle that for you

tyler-cloud-elements replied:
As you're only using `jpgFile` here, you can just inline it, i.e. 
```
`${__dirname}/assets/brady.jpg`
```

tyler-cloud-elements commented on src/test/elements/economic/journals.js:
```diff
@@ -0,0 +1,46 @@
+     })
+     .should.return200OnGet();
+
+     it.skip('should support S for /journals/:id/entires, /journals/:id/vouchers and CSD for /journals/:id/accountingYear:year/vouchers/id/attachments  ', () => {
```
Remove spaces here

tyler-cloud-elements commented on src/test/elements/economic/payment-terms.js:
```diff
@@ -0,0 +1,23 @@
+const suite = require('core/suite');
+const tools = require('core/tools');
+const chakram = require('chakram');
+const expect = require('chakram').expect;
```
Can just do `const expect = chakram.expect;` here

tyler-cloud-elements commented on src/test/elements/economic/product-groups.js:
```diff
@@ -0,0 +1,30 @@
+       return cloud.get(`${test.api}`)
+         .then(r => {
+           id = r.body[0].id;
+         })
```
In line plz

tyler-cloud-elements commented on src/test/elements/economic/product-groups.js:
```diff
@@ -0,0 +1,30 @@
+       expect(validValues.length).to.equal(r.body.length);
+     })
+     .should.return200OnGet();
+     it('should support S for /product-groups/:id/sales-accounts', () => {
```
Format plz

MadhuriKadam-GSLab replied:
done

tyler-cloud-elements commented on src/test/elements/economic/products.js:
```diff
@@ -0,0 +1,23 @@
+const suite = require('core/suite');
+const tools = require('core/tools');
+const chakram = require('chakram');
+const expect = require('chakram').expect;
```
Same as above

MadhuriKadam-GSLab replied:
done

tyler-cloud-elements commented on src/test/elements/economic/vat-types.js:
```diff
@@ -0,0 +1,29 @@
+       return cloud.get(`${test.api}`)
+         .then(r => {
+           Vid = r.body[0].id;
+         })
```
Inline plz

tyler-cloud-elements commented on src/test/elements/economic/vat-types.js:
```diff
@@ -0,0 +1,29 @@
+     })
+     .should.return200OnGet();
+
+     it('should support SR for /vat-types/:id/vat-report-setups', () => {
```
format

tyler-cloud-elements commented on src/test/elements/economic/vendors.js:
```diff
@@ -0,0 +1,23 @@
+const payload = tools.requirePayload(`${__dirname}/assets/vendors.json`);
+
+suite.forElement('erp', 'vendors', { payload: payload }, (test) => {
+  payload.supplierNumber=tools.randomInt();
```
Add next to the `const payload`

MadhuriKadam-GSLab commented:
@tyler-cloud-elements  I have updated review comments. Updated authentication flow to oauth2 hence need to update the churros for authentication will update that tomorrow  along with review  comments
